SPEC.rdoc

```diff
@@ -138,6 +138,9 @@ There are the following restrictions:
   <tt>SCRIPT_NAME</tt> is empty.
   <tt>SCRIPT_NAME</tt> never should be <tt>/</tt>, but instead be empty.
 <tt>rack.response_finished</tt>:: An array of callables run after the HTTP response has been sent.
+The array of callables should be called directly after the HTTP response has
+been closed. The callables should be called sequentially and synchronously.
```

Comment:
This is great, but I'd suggest some minor terminology adjustments:

```suggestion
The array of callables should be called directly after the HTTP response has
been sent. The callables should be called sequentially and synchronously in
the same execution context as the response. If an exception is raised, it will
be ignored and will not impact the execution of other callables.
```

I'm not strong on the exception handling clause, the alternative is to propagate it up and fail the response (i.e. log).

Reply:
A quick survey:

## Puma

Doesn't look like there is any error handling:

https://github.com/puma/puma/blob/master/lib/puma/request.rb#L185

Based on the caller, it looks like the entire connection will be closed:

https://github.com/puma/puma/blob/db751ba82f8980bc3095fc01cbd4458480b804fc/lib/puma/server.rb#L445

## Unicorn

Doesn't look like any error handling:

https://github.com/defunkt/unicorn/blob/master/lib/unicorn/http_server.rb#L657

Based on the caller it looks like it would take out the entire connection:

https://github.com/defunkt/unicorn/blob/a08e4a6d613d9f3b4c2afb57f2691f0710a9d920/lib/unicorn/http_server.rb#L732-L751

Reply:
I think we need to consider the safety trade off that is being made by the current implementation of `after_reply`. I can see three approaches:

1. The current approach - do nothing and the entire connection might be trashed.
2. `begin...rescue` around ALL callables.
3. `begin...rescue` around EACH callable.

1 is most efficient / least safe <-> 3 is least efficient / most safe.

Reply:
Testing this out:

```
run ->(env) do
  env['rack.after_reply'] << lambda do
    raise "Fail"
  end

  [200, [], []]
end
```

```
> puma 
Puma starting in single mode...
* Puma version: 5.6.2 (ruby 3.1.1-p18) ("Birdie's Version")
*  Min threads: 0
*  Max threads: 5
*  Environment: development
*          PID: 19925
* Listening on http://0.0.0.0:9292
Use Ctrl-C to stop
2022-04-09 11:57:26 +1200 Read: #<RuntimeError: Fail>
```

Puma logs the error, but:

```
> curl -v http://localhost:9292 http://localhost:9292
*   Trying 127.0.0.1:9292...
* Connected to localhost (127.0.0.1) port 9292 (#0)
> GET / HTTP/1.1
> Host: localhost:9292
> User-Agent: curl/7.80.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Transfer-Encoding: chunked
< 
* Leftovers after chunking:  38u bytes
* Connection #0 to host localhost left intact
* Found bundle for host localhost: 0x6000036d9410 [serially]
* Connection 0 seems to be dead!
* Closing connection 0
* Hostname localhost was found in DNS cache
*   Trying 127.0.0.1:9292...
* Connected to localhost (127.0.0.1) port 9292 (#1)
> GET / HTTP/1.1
> Host: localhost:9292
> User-Agent: curl/7.80.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Transfer-Encoding: chunked
< 
* Leftovers after chunking:  38u bytes
* Connection #1 to host localhost left intact
```

Note that:

```
* Connection #0 to host localhost left intact
* Found bundle for host localhost: 0x6000036d9410 [serially]
* Connection 0 seems to be dead!
```

which indicates the connection has failed internally.

I suspect something similar for Unicorn.

Reply:
> I think we need to consider the safety trade off that is being made by the current implementation of `after_reply`. I can see three approaches:
> 
> 1. The current approach - do nothing and the entire connection might be trashed.
> 2. `begin...rescue` around ALL callables.
> 3. `begin...rescue` around EACH callable.
> 
> 1 is most efficient / least safe <-> 3 is least efficient / most safe.

If we go with 2 or 3 that might be a good reason to change the name of the feature since the existing implementations won't be compatible with the spec.

I view this as a low-level API, so I think the existing behavior is likely fine. Internally we're using a small wrapper on-top of `rack.after_reply` to handle exceptions, reporting, metrics, etc. There's also [maybe_later](https://github.com/testdouble/maybe_later) which is a gem that implements similar logic.

I'm not familiar with how Puma handles dead connections, will it restart the connection or replace it so there's no impact on the running application?

Reply:
Agree that the existing behavior is fine as this is a low-level method. This still needs a decision on naming from @tenderlove (or another committer).

I don't think option 2 makes sense.  What would a rescue around all callables do?  Maybe just allow skipping the webserver error logging?

A rescue around each callable (option 3) has some utility, in that you can add your callback and be fairly sure it will run even if another callback fails.  

Note that if want to standardize different behavior (that the server must rescue all/each callable), then I don't have an objection to changing the `env` key.  In that case, it would be best to pick a `env` key that wasn't already used, so that we don't make existing webservers violate the spec.

Reply:
> A rescue around each callable (option 3) has some utility, in that you can add your callback and be fairly sure it will run even if another callback fails.

I personally think this is the safest option.

Regarding option 2, the reality is servers don't do anything at all and it looks to be breaking (e.g. Puma) internal request handling. So at least stating that; "If a callable fails, it won't break the entire connection" is something we should be clear on in the spec IMHO if we opt to go in that direction.

Reply:
> I'm not familiar with how Puma handles dead connections, will it restart the connection or replace it so there's no impact on the running application?

It looks like a server bug as the error is totally unexpected and basically seems to cause strange handling of the connection internally.

Reply:
> Regarding option 2, the reality is servers don't do anything at all and it looks to be breaking (e.g. Puma) internal request handling. So at least stating that; "If a callable fails, it won't break the entire connection" is something we should be clear on in the spec IMHO if we opt to go in that direction.

I agree. I think if we're going in that direction providing a guarantee that your `callable` will run is a logical next step to that.

I could imagine there being some cases where you'd want to stop the callable chain early, specifically since it's taking up worker time but I think building abstractions like the `maybe_later` gem will largely cover that. The question I have is where should guarantees about each callable running live?

* Should we define the spec to declare that `callable`s will all be run, even if previous `callable`s fail and rely on abstractions like `maybe_later` to enable short-circuiting running `callable`s?
* Should we define the spec to declare that `callable`s that raise won't break the server, and rely on abstractions like `maybe_later` to enable guarantees that each of your callables will run?

Alternatively, we could go with #3 (the first option above) and have a special error class that when raised will prevent subsequent callables from running.

Reply:
> Alternatively, we could go with #3 (the first option above) and have a special error class that when raised will prevent subsequent callables from running.

All of the rack SPEC is designed so you do not need to include rack the library.  So an exception class defined by rack the library would not work.  One way around this would be to say you can throw `:skip_response_finished`  inside a `rack.response_finished` entry to skip the remaining entries.  This is easy for servers to implement via `catch` around the code that calls each of the entries.

Reply:
> One way around this would be to say you can throw `:skip_response_finished` inside a `rack.response_finished` entry to skip the remaining entries. This is easy for servers to implement via `catch` around the code that calls each of the entries.

Ah I see, thank you for the context. 

That general approach makes sense to me. Maybe instead of throwing, we could rely on returning `:skip_response_finished` from the `callable`?

Reply:
I don't personally believe that there is any good use case where one callable should be able to cancel subsequent callables as this seems order dependent and pretty fragile.

I'd be okay with exceptions terminating all remaining callables as a performance/safety middle ground.